pkg/database/all_ops.go

```diff
@@ -82,29 +88,39 @@ func (d *db) ExecAll(req *schema.ExecAllRequest) (*schema.TxHeader, error) {
 					return nil, store.ErrIllegalArguments
 				}
 
-				// check key does not exists or it's already a reference
-				entry, err := d.getAt(EncodeKey(x.Ref.Key), 0, 0, index, tx)
-				if err != nil && err != store.ErrKeyNotFound {
-					return nil, err
-				}
-				if entry != nil && entry.ReferencedBy == nil {
-					return nil, ErrFinalKeyCannotBeConvertedIntoReference
+				if req.NoWait && (x.Ref.AtTx != 0 || !x.Ref.BoundRef) {
+					return nil, ErrNoWaitOperationMustBeSelfContained
 				}
 
-				// reference arguments are converted in regular key value items and then atomically inserted
 				_, exists := kmap[sha256.Sum256(x.Ref.ReferencedKey)]
 
-				if !exists || x.Ref.AtTx > 0 {
-					// check referenced key exists and it's not a reference
-					refEntry, err := d.getAt(EncodeKey(x.Ref.ReferencedKey), x.Ref.AtTx, 0, index, tx)
-					if err != nil {
+				if req.NoWait && !exists {
+					return nil, ErrNoWaitOperationMustBeSelfContained
+				}
```

Comment:
We could also check here if the referenced entry from current transaction is not a reference itself,

also we could check if there was not previous entry with the same key name in the same ExecAll and that it was not changed from value to a reference with the current operation (or to restrict to only one value in whole ExecAll for one key but this could break backwards compatibility with some odd applications).